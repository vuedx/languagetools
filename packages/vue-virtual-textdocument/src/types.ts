export const TEMPLATE_BLOCK_SELECTOR = 'template'
export const SCRIPT_BLOCK_SELECTOR = 'script'
export const SCRIPT_SETUP_BLOCK_SELECTOR = 'scriptSetup'
export const RENDER_SELECTOR = '_render'
export const MODULE_SELECTOR = '_module'
export const INTERNAL_MODULE_SELECTOR = '_internal'
export const TEMPLATE_AST_SELECTOR = '_ast'

export type BlockSelector =
  | { type: typeof SCRIPT_BLOCK_SELECTOR }
  | { type: typeof SCRIPT_SETUP_BLOCK_SELECTOR }
  | { type: typeof TEMPLATE_BLOCK_SELECTOR }
  | { type: 'style'; index: number }
  | { type: 'customBlocks'; index: number }

export type Selector =
  | BlockSelector
  | { type: typeof RENDER_SELECTOR }
  | { type: typeof MODULE_SELECTOR }
  | { type: typeof INTERNAL_MODULE_SELECTOR }
  | { type: typeof TEMPLATE_AST_SELECTOR }

export type SelectorLike =
  | Selector
  | typeof TEMPLATE_BLOCK_SELECTOR
  | typeof SCRIPT_BLOCK_SELECTOR
  | typeof SCRIPT_SETUP_BLOCK_SELECTOR
  | typeof RENDER_SELECTOR
  | typeof MODULE_SELECTOR
  | typeof TEMPLATE_AST_SELECTOR
  | typeof INTERNAL_MODULE_SELECTOR
