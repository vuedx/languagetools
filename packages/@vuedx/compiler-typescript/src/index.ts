import { parse } from '@babel/parser';
import {
  ExportNamedDeclaration,
  FunctionDeclaration,
  isExportNamedDeclaration,
  isReturnStatement,
  ReturnStatement,
} from '@babel/types';
import { baseParse, transform, CodegenResult } from '@vue/compiler-core';
import { compile as baseCompile } from './compile';
import { match } from './matcher';
import { transformExpression as transformExpressionAlternate } from './nodes/transformExpression';
import { Options } from './options';

export * from '@vue/compiler-core';
export * from './mapper';
export * from './runtimeHelpers';
export * from './interfaces';

export function compile(source: string, options: Options): CodegenResult {
  const ast = baseParse(source, options);
  const { code, map } = baseCompile(clone(ast), options);
  const renderAST = parse(code, {
    ranges: true,
    sourceType: 'module',
    plugins: options.useJavaScript ? [] : ['typescript'],
  });
  const renderFn: FunctionDeclaration = (renderAST.program.body.find((statement) =>
    isExportNamedDeclaration(statement)
  ) as ExportNamedDeclaration).declaration;
  const generated = (renderFn.body.body.find((statement) => isReturnStatement(statement)) as ReturnStatement).argument!;

  transform(ast, {
    nodeTransforms: [transformExpressionAlternate],
  });

  match(ast, null as any, generated);

  return { code, ast, map };
}

function clone<T>(object: T): T {
  return JSON.parse(JSON.stringify(object)) as T;
}
