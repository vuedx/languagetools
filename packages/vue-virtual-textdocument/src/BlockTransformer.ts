import type {
  CompilerError,
  RawSourceMap,
  SFCBlock,
  SFCDescriptor,
  SFCScriptBlock,
  SFCStyleBlock,
  SFCTemplateBlock,
} from '@vuedx/compiler-sfc'
import { annotations, compile } from '@vuedx/compiler-tsx'
import { pascalCase } from '@vuedx/shared'
import { isRootNode } from '@vuedx/template-ast-types'
import {
  createExportDeclarationForComponents,
  createExportDeclarationForDirectives,
  createExportDeclarationForExpose,
  createExportDeclarationForComponent,
  findScopeBindings,
  toAST,
  toCode,
} from '@vuedx/transforms'
import * as Path from 'path'
import type { VueSFCDocument } from './VueSFCDocument'

export { annotations }
export type Annotations = typeof annotations

export interface BlockTransformerOptions {
  block: SFCBlock | SFCScriptBlock | SFCStyleBlock | SFCTemplateBlock
  document: VueSFCDocument
  descriptor: SFCDescriptor
  annotations: Annotations
}

export interface TransformerError {
  message: string
  code: number | string | undefined
  codeDescription?: {
    href: string
  }
  start: number | undefined
  length: number | undefined
  source?: string
  severity: 'error' | 'warning' | 'info' | 'hint' | 'unused' | 'deprecated'
}

export type BlockTransformerFn = (
  source: string,
  id: string,
  options: BlockTransformerOptions,
) => {
  code: string
  map?: RawSourceMap
  errors?: TransformerError[]
  ast?: any
}

export interface BlockTransformer {
  transform: BlockTransformerFn
  output(block: SFCBlock): 'js' | 'jsx' | 'ts' | 'tsx'
}

function isSupportedLang(lang?: string): lang is 'js' | 'jsx' | 'ts' | 'tsx' {
  return ['js', 'jsx', 'ts', 'tsx'].includes(lang ?? '')
}

const caches = {
  script: new WeakMap<VueSFCDocument, string>(),
}

function toText(...lines: string[]): string {
  return lines.join('\n').replace(/\n$/, '') + '\n'
}

export const builtins: Record<'script' | 'template', BlockTransformer> = {
  script: {
    output: (block) => (isSupportedLang(block.lang) ? block.lang : 'js'),
    transform: (source, _id, { block, document }) => {
      try {
        const isScriptSetup =
          'setup' in block && block.setup != null && block.setup !== false
        // TODO: Should it be typescript instead of babel?
        const ast = toAST(source, {
          sourceFilename: document.fileName,
          isScriptSetup: isScriptSetup,
          lang: block.lang,
        })

        const usedIdentifiers = new Set<string>()
        const usedDirectives = new Set<string>()
        const usedComponents = new Set<string>()

        if (document.descriptor.template != null && isScriptSetup) {
          const ast = document.getDoc(document.descriptor.template)?.ast

          if (isRootNode(ast)) {
            ast.scope.globals.forEach((id) => {
              usedIdentifiers.add(id)
            })

            ast.components.forEach((name) => {
              const id = pascalCase(name.split('.').shift() ?? name)
              usedComponents.add(id)
            })

            ast.directives.forEach((name) => {
              const id = 'v' + pascalCase(name)
              usedDirectives.add(id)
            })
          }
        }

        const config = {
          isScriptSetup,
          leadingCommentForCopiedSource: annotations.copiedSource.start,
          trailingCommentForCopiedSource: annotations.copiedSource.end,
          leadingCommentForIdentifiers: annotations.templateGlobals.start,
          trailingCommentForIdentifiers: annotations.templateGlobals.end,
        }

        const nodes: any[] = [
          createExportDeclarationForComponents(ast, {
            ...config,
            shouldIncludeScriptSetup: (id) => usedComponents.has(id),
          }),
          createExportDeclarationForDirectives(ast, {
            ...config,
            shouldIncludeScriptSetup: (id) => usedDirectives.has(id),
          }),
          createExportDeclarationForExpose(ast, config),
        ]

        if (isScriptSetup) {
          document.declarations.identifiers = new Set(findScopeBindings(ast))

          nodes.push(
            createExportDeclarationForComponent(ast, {
              ...config,
              shouldIncludeScriptSetup: (id) => usedIdentifiers.has(id),
            }),
          )
        } else {
          nodes.push(createExportDeclarationForComponent(ast, config))
        }

        const code = toText(
          ';',
          annotations.diagnosticsIgnore.start,
          ';' + annotations.tsCompletions,
          toCode(nodes, { sourceText: source }).code,
          annotations.diagnosticsIgnore.end,
        )

        caches.script.set(document, code)

        return {
          ast,
          code: source + code,
        }
      } catch (error) {
        console.error('[VueDX] BlockTransformer - Error parsing script:', error)

        return {
          code:
            source +
            (caches.script.get(document) ??
              toText(
                ';',
                annotations.diagnosticsIgnore.start,
                ';' + annotations.tsCompletions,
                'export const __VueDX_components = {};',
                'export const __VueDX_directives = {};',
                'export const __VueDX_expose = {};',
                'export const __VueDX_DefineComponent = VueDX.internal.defineComponent({});',
                annotations.diagnosticsIgnore.end,
              )),
          errors: [],
        }
      }
    },
  },
  template: {
    output: () => 'tsx',
    transform: (source, _, { document, descriptor, block }) => {
      try {
        const selfSrcFileName = document.getBlockId(
          descriptor.scriptSetup ??
            descriptor.script ??
            document.fallbackScript,
        )

        const errors: CompilerError[] = []
        const result = compile(source, {
          sourceMap: true,
          filename: document.fileName,

          selfSrc: `./${Path.posix.basename(
            selfSrcFileName.substring(
              0,
              selfSrcFileName.length -
                Path.posix.extname(selfSrcFileName).length,
            ),
          )}`,

          onError(_e) {
            // errors.push(error)
          },
        })

        errors.push(...result.errors)
        document.templateAST = result.ast

        return {
          code: result.code,
          map: result.map,
          ast: result.ast,
          errors: errors.map((error) => {
            const diag: TransformerError = {
              message: error.message,
              code: error.code,
              severity: 'error',
              source: 'TemplateCompiler',
              start: undefined,
              length: undefined,
            }

            if (error.loc != null) {
              diag.start = error.loc.start.offset
              diag.length = error.loc.end.offset - error.loc.start.offset
            }

            return diag
          }),
        }
      } catch (e) {
        const error = e as Error
        return {
          code: '',

          errors: [
            {
              message: `Error processing template: ${String(
                error.message,
              )}\n${String(error.stack)}`,
              code: -1,
              source: 'TemplateCompiler',
              severity: 'error',
              start: 0,
              length: block.loc.source.length,
            },
          ],
        }
      }
    },
  },
}
