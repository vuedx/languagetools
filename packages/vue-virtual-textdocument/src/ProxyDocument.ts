import type {
  Position,
  Range,
  TextDocument,
} from 'vscode-languageserver-textdocument'

export abstract class ProxyDocument implements TextDocument {
  constructor(protected source: TextDocument) {}

  public get uri(): string {
    return this.source.uri
  }

  public get languageId(): string {
    return this.source.languageId
  }

  public get version(): number {
    return this.source.version
  }

  public get lineCount(): number {
    return this.source.lineCount
  }

  public getText(range?: Range): string {
    return this.source.getText(range)
  }

  public positionAt(offset: number): Position {
    return this.source.positionAt(offset)
  }

  public offsetAt(position: Position): number {
    return this.source.offsetAt(position)
  }
}
