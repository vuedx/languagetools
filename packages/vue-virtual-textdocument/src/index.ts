export {
  Position,
  Range,
  TextDocument,
  TextDocumentContentChangeEvent,
} from 'vscode-languageserver-textdocument'
export * from './VueSFCDocument'
