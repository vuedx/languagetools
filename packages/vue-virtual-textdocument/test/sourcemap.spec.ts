import { TextSpan, VueSFCDocument } from '../src'

expect.addSnapshotSerializer({
  serialize(val) {
    return String(val)
  },
  test(val) {
    return typeof val === 'string'
  },
})

describe('sourcemaps', () => {
  beforeEach(() => {
    jest.spyOn(console, 'debug').mockImplementation(() => {})
  })

  test(`directives`, () => {
    const code = `
    <template>
      <B a="" @a="onNum" @b.once="onNum" @c="call.a['func']($event)" />
    </template>
    `
    const file = getTemplateFile(code)
    const first = code.indexOf('onNum')
    const second = code.indexOf('onNum', first)
    const once = code.indexOf('once')
    const call = code.indexOf('call.a')
    const event = code.indexOf('$event')

    const spans: TextSpan[] = [
      { start: first, length: 5 },
      { start: first + 3, length: 2 },
      { start: second, length: 5 },
      { start: second + 3, length: 2 },
      { start: once, length: 4 },
      { start: call + 6, length: 8 },
      { start: event, length: 6 },
    ]

    spans.forEach((span) => {
      checkGeneratedSpan(file, span)
    })
  })

  test(`v-text`, () => {
    const code = `
    <template>
      <div v-text="a" />
    </template>
    `

    const file = getTemplateFile(code)
    const offset = code.indexOf('"a"') + 1

    checkGeneratedSpan(file, { start: offset, length: 1 })
  })
  test.only(`v-on shorthand`, () => {
    const code = `
    <template>
      <B a="" @a="onStr" />
    </template>
    `

    const file = getTemplateFile(code)
    const offset = code.indexOf('@a') + 1

    checkGeneratedSpan(file, { start: offset, length: 1 })
  })

  test(`component`, () => {
    const code = `
<script setup lang="ts">
import { ref } from 'vue'

const a = ref(false)
let b  = 5
let c: { foo: string} | undefined
</script>

<template>
  <div v-text="a" />
  <div v-text="b" />
  <div v-text="c" />
  <div v-text="c?.foo" />
</template>
    `
    const file = getTemplateFile(code)

    expect(file.doc.generated.getText()).toMatchSnapshot()
  })
})

function checkGeneratedSpan(
  file: ReturnType<typeof getTemplateFile>,
  original: TextSpan,
) {
  const generated = file.doc.findGeneratedTextSpan(original)
  if (generated == null) throw new Error('Generated span is null')
  checkSnapshot(file, original, generated)
  const original2 = file.doc.findOriginalTextSpan(generated)
  if (original2 == null) throw new Error('Original span is null')
  checkSnapshot(file, original2, generated)
}

function checkSnapshot(
  file: ReturnType<typeof getTemplateFile>,
  original: TextSpan,
  generated: TextSpan,
) {
  const first = file.snapshot(original, 2, 'original')
  const second = file.snapshot(generated, 2, 'generated')
  expect(`${first}\n----\n${second}`).toMatchSnapshot()
}

function getTemplateFile(code: string) {
  const file = VueSFCDocument.create('/foo/bar/Example.vue', code)

  function original(range: TextSpan) {
    return file.original
      .getText()
      .slice(range.start, range.start + range.length)
  }

  function generated(range: TextSpan) {
    return file.generated
      .getText()
      .slice(range.start, range.start + range.length)
  }

  function snapshot(
    range: TextSpan | null,
    size = 2,
    kind: 'generated' | 'original' = 'generated',
  ) {
    if (range == null) throw new Error('Range is null')
    const source = kind === 'generated' ? file.generated : file.original
    const text = source.getText()
    const line = (start: number) => {
      const index = text.indexOf('\n', start)
      if (index >= 0) return text.substring(start, index)
      return text.substring(start)
    }
    const { line: L, character: C } = source.positionAt(range.start)
    const lines: string[] = []
    for (let i = 1; i <= size && L - i >= 0; ++i) {
      const start = source.offsetAt({ line: L - i, character: 0 })
      lines.unshift(line(start))
    }

    const start = source.offsetAt({ line: L, character: 0 })
    lines.push(line(start))
    lines.push(
      ' '.repeat(C) +
        '^'.repeat(range.length) +
        ` > ${range.length} at ${range.start} (${L}:${C})`,
    )

    for (let i = 1; i <= size && L + i < source.lineCount; ++i) {
      const start = source.offsetAt({ line: L + i, character: 0 })
      lines.push(line(start))
    }

    return lines.join('\n')
  }

  return { doc: file, original, generated, snapshot }
}
