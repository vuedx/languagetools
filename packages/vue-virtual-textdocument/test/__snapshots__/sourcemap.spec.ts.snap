// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sourcemaps attrs types 1`] = `
import * as __VueDX_TypeCheck from 'vuedx~runtime';
import { defineComponent as __VueDX_defineComponent, GlobalComponents as __VueDX_GlobalComponents } from 'vue';
//#region <script>

const __VueDX__Script_Component = __VueDX_defineComponent({});
//#endregion
//#region <script setup>
 ;const __VueDX__ScriptSetup_scope = __VueDX_TypeCheck.internal.scope(async () => {
const __VueDX__ScriptSetup_props = defineProps({});
const __VueDX__ScriptSetup_emits = ({});
const __VueDX__ScriptSetup_expose = {};
const __VueDX__ScriptSetup_internalProps = {};
const __VueDX__ScriptSetup_Component = __VueDX_defineComponent((_: typeof __VueDX__ScriptSetup_internalProps)=> {});

//#endregion
function __VueDX_RegisterSelf<T>(ctx: T) {
  return { ...ctx, ["Example"]: Example }
}
const __VueDX_ctx = __VueDX_RegisterSelf(new __VueDX__Script_Component())
//#region <template>
function __VueDX_render() {
  /*<vuedx:templateGlobals>*/
  let a = __VueDX_ctx.a
  let b = __VueDX_ctx.b
  let c = __VueDX_ctx.c
  let $slots = __VueDX_ctx.$slots
  /*</vuedx:templateGlobals>*/
  return (
    <>
      {a
        ? <>
            <div
              /*<vuedx:tsx-completions-target/>*/
            />
          </>
        : b
        ? <>
            <span
              /*<vuedx:tsx-completions-target/>*/
            />
          </>
        : <>
            {c
              ? <>
                  <p
                    /*<vuedx:tsx-completions-target/>*/
                  />
                </>
              : null
            }
          </>
      }
    </>
  )
}
__VueDX_render();
/*<vuedx:diagnosticsIgnore>*/
function __VueDX__slots() {
  /*<vuedx:templateGlobals>*/
  let a = __VueDX_ctx.a
  let b = __VueDX_ctx.b
  let c = __VueDX_ctx.c
  let $slots = __VueDX_ctx.$slots
  /*</vuedx:templateGlobals>*/
  return __VueDX_TypeCheck.internal.union(__VueDX_TypeCheck.internal.flat([
  ]))
}
/*</vuedx:diagnosticsIgnore>*/
const __VueDX__attrs = (() => {
  return __VueDX_TypeCheck.internal.first(__VueDX_TypeCheck.internal.flat([
    {} as unknown as __VueDX_TypeCheck.internal.PropsOf<JSX.IntrinsicElements, "div">,
    {} as unknown as __VueDX_TypeCheck.internal.PropsOf<JSX.IntrinsicElements, "span">,
    {} as unknown as __VueDX_TypeCheck.internal.PropsOf<JSX.IntrinsicElements, "p">,
  ]))
})();
//#endregion
return {__VueDX__ScriptSetup_Component, __VueDX__ScriptSetup_props, __VueDX__ScriptSetup_emits, __VueDX__ScriptSetup_expose, __VueDX__attrs, __VueDX__slots, __VueDX_ctx};});
const {__VueDX__ScriptSetup_Component, __VueDX__ScriptSetup_props, __VueDX__ScriptSetup_emits, __VueDX__ScriptSetup_expose, __VueDX__attrs, __VueDX__slots, __VueDX_ctx} = __VueDX__ScriptSetup_scope;
//#region public component definition
const ExamplePublic = null as unknown as new () => typeof __VueDX__ScriptSetup_expose;
export default class Example extends ExamplePublic {
  $props = null as unknown as __VueDX_TypeCheck.internal.MergeAttrs<typeof __VueDX_ctx.$props, typeof __VueDX__attrs>;
  $slots = null as unknown as __VueDX_TypeCheck.internal.Slots<ReturnType<typeof __VueDX__slots>>;
}
//#endregion

`;

exports[`sourcemaps component 1`] = `
import * as __VueDX_TypeCheck from 'vuedx~runtime';
import { defineComponent as __VueDX_defineComponent, GlobalComponents as __VueDX_GlobalComponents } from 'vue';
//#region <script>

const __VueDX__Script_Component = __VueDX_defineComponent({});
//#endregion
//#region <script setup>

import { ref } from 'vue' ;const __VueDX__ScriptSetup_scope = __VueDX_TypeCheck.internal.scope(async () => {

const a = ref(false)
let b  = 5
let c: { foo: string} | undefined
const __VueDX__ScriptSetup_props = defineProps({});
const __VueDX__ScriptSetup_emits = ({});
const __VueDX__ScriptSetup_expose = {};
const __VueDX__ScriptSetup_internalProps = {};
const __VueDX__ScriptSetup_Component = __VueDX_defineComponent((_: typeof __VueDX__ScriptSetup_internalProps)=> {});

//#endregion
function __VueDX_RegisterSelf<T>(ctx: T) {
  return { ...ctx, ["Example"]: Example }
}
const __VueDX_ctx = __VueDX_RegisterSelf(new __VueDX__ScriptSetup_Component())
//#region <template>
/*<vuedx:templateGlobals>*/
const __VueDX__get_identifier_a = () => __VueDX_TypeCheck.internal.unref(a);
const __VueDX__get_identifier_b = () => __VueDX_TypeCheck.internal.unref(b);
const __VueDX__get_identifier_c = () => __VueDX_TypeCheck.internal.unref(c);
/*</vuedx:templateGlobals>*/
function __VueDX_render() {
  /*<vuedx:templateGlobals>*/
  let a = __VueDX__get_identifier_a();
  let b = __VueDX__get_identifier_b();
  let c = __VueDX__get_identifier_c();
  let $slots = __VueDX_ctx.$slots
  /*</vuedx:templateGlobals>*/
  return (
    <>
      <div
        innerHTML={a}
        /*<vuedx:tsx-completions-target/>*/
      />
      <div
        innerHTML={b}
        /*<vuedx:tsx-completions-target/>*/
      />
      <div
        innerHTML={c}
        /*<vuedx:tsx-completions-target/>*/
      />
      <div
        innerHTML={c?.foo}
        /*<vuedx:tsx-completions-target/>*/
      />
    </>
  )
}
__VueDX_render();
/*<vuedx:diagnosticsIgnore>*/
function __VueDX__slots() {
  /*<vuedx:templateGlobals>*/
  let a = __VueDX__get_identifier_a();
  let b = __VueDX__get_identifier_b();
  let c = __VueDX__get_identifier_c();
  let $slots = __VueDX_ctx.$slots
  /*</vuedx:templateGlobals>*/
  return __VueDX_TypeCheck.internal.union(__VueDX_TypeCheck.internal.flat([
  ]))
}
/*</vuedx:diagnosticsIgnore>*/
const __VueDX__attrs = (() => {
  return __VueDX_TypeCheck.internal.first(__VueDX_TypeCheck.internal.flat([{}]))
})();
//#endregion
return {__VueDX__ScriptSetup_Component, __VueDX__ScriptSetup_props, __VueDX__ScriptSetup_emits, __VueDX__ScriptSetup_expose, __VueDX__attrs, __VueDX__slots, __VueDX_ctx};});
const {__VueDX__ScriptSetup_Component, __VueDX__ScriptSetup_props, __VueDX__ScriptSetup_emits, __VueDX__ScriptSetup_expose, __VueDX__attrs, __VueDX__slots, __VueDX_ctx} = __VueDX__ScriptSetup_scope;
//#region public component definition
const ExamplePublic = null as unknown as new () => typeof __VueDX__ScriptSetup_expose;
export default class Example extends ExamplePublic {
  $props = null as unknown as __VueDX_TypeCheck.internal.MergeAttrs<typeof __VueDX_ctx.$props & __VueDX_TypeCheck.internal.EmitsToProps<typeof __VueDX__ScriptSetup_emits>, typeof __VueDX__attrs>;
  $slots = null as unknown as __VueDX_TypeCheck.internal.Slots<ReturnType<typeof __VueDX__slots>>;
}
//#endregion

`;

exports[`sourcemaps directives 1`] = `

    <template>
      <B a="" @a="onNum" @b.once="onNum" @c="call.a['func']($event)" />
                  ^^^^^ > 5 at 34 (2:18)
    </template>
    
----
        a=""
        onA={__VueDX_TypeCheck.internal.first([
          onNum, 
          ^^^^^ > 5 at 1516 (37:10)
        ])}
        onB={__VueDX_TypeCheck.internal.first([
`;

exports[`sourcemaps directives 2`] = `

    <template>
      <B a="" @a="onNum" @b.once="onNum" @c="call.a['func']($event)" />
                  ^^^^^ > 5 at 34 (2:18)
    </template>
    
----
        a=""
        onA={__VueDX_TypeCheck.internal.first([
          onNum, 
          ^^^^^ > 5 at 1516 (37:10)
        ])}
        onB={__VueDX_TypeCheck.internal.first([
`;

exports[`sourcemaps directives 3`] = `

    <template>
      <B a="" @a="onNum" @b.once="onNum" @c="call.a['func']($event)" />
                     ^^ > 2 at 37 (2:21)
    </template>
    
----
        a=""
        onA={__VueDX_TypeCheck.internal.first([
          onNum, 
             ^^ > 2 at 1519 (37:13)
        ])}
        onB={__VueDX_TypeCheck.internal.first([
`;

exports[`sourcemaps directives 4`] = `

    <template>
      <B a="" @a="onNum" @b.once="onNum" @c="call.a['func']($event)" />
                     ^^ > 2 at 37 (2:21)
    </template>
    
----
        a=""
        onA={__VueDX_TypeCheck.internal.first([
          onNum, 
             ^^ > 2 at 1519 (37:13)
        ])}
        onB={__VueDX_TypeCheck.internal.first([
`;

exports[`sourcemaps directives 5`] = `

    <template>
      <B a="" @a="onNum" @b.once="onNum" @c="call.a['func']($event)" />
                  ^^^^^ > 5 at 34 (2:18)
    </template>
    
----
        a=""
        onA={__VueDX_TypeCheck.internal.first([
          onNum, 
          ^^^^^ > 5 at 1516 (37:10)
        ])}
        onB={__VueDX_TypeCheck.internal.first([
`;

exports[`sourcemaps directives 6`] = `

    <template>
      <B a="" @a="onNum" @b.once="onNum" @c="call.a['func']($event)" />
                  ^^^^^ > 5 at 34 (2:18)
    </template>
    
----
        a=""
        onA={__VueDX_TypeCheck.internal.first([
          onNum, 
          ^^^^^ > 5 at 1516 (37:10)
        ])}
        onB={__VueDX_TypeCheck.internal.first([
`;

exports[`sourcemaps directives 7`] = `

    <template>
      <B a="" @a="onNum" @b.once="onNum" @c="call.a['func']($event)" />
                     ^^ > 2 at 37 (2:21)
    </template>
    
----
        a=""
        onA={__VueDX_TypeCheck.internal.first([
          onNum, 
             ^^ > 2 at 1519 (37:13)
        ])}
        onB={__VueDX_TypeCheck.internal.first([
`;

exports[`sourcemaps directives 8`] = `

    <template>
      <B a="" @a="onNum" @b.once="onNum" @c="call.a['func']($event)" />
                     ^^ > 2 at 37 (2:21)
    </template>
    
----
        a=""
        onA={__VueDX_TypeCheck.internal.first([
          onNum, 
             ^^ > 2 at 1519 (37:13)
        ])}
        onB={__VueDX_TypeCheck.internal.first([
`;

exports[`sourcemaps directives 9`] = `

    <template>
      <B a="" @a="onNum" @b.once="onNum" @c="call.a['func']($event)" />
                                                   ^^^^^^^^ > 8 at 67 (2:51)
    </template>
    
----
        onC={__VueDX_TypeCheck.internal.first([
          ($event) => {
          call.a['func']($event)
                ^^^^^^^^ > 8 at 1702 (44:16)
          }, 
        ])}
`;

exports[`sourcemaps directives 10`] = `

    <template>
      <B a="" @a="onNum" @b.once="onNum" @c="call.a['func']($event)" />
                                                   ^^^^^^^^ > 8 at 67 (2:51)
    </template>
    
----
        onC={__VueDX_TypeCheck.internal.first([
          ($event) => {
          call.a['func']($event)
                ^^^^^^^^ > 8 at 1702 (44:16)
          }, 
        ])}
`;

exports[`sourcemaps directives 11`] = `

    <template>
      <B a="" @a="onNum" @b.once="onNum" @c="call.a['func']($event)" />
                                                            ^^^^^^ > 6 at 76 (2:60)
    </template>
    
----
        onC={__VueDX_TypeCheck.internal.first([
          ($event) => {
          call.a['func']($event)
                         ^^^^^^ > 6 at 1711 (44:25)
          }, 
        ])}
`;

exports[`sourcemaps directives 12`] = `

    <template>
      <B a="" @a="onNum" @b.once="onNum" @c="call.a['func']($event)" />
                                                            ^^^^^^ > 6 at 76 (2:60)
    </template>
    
----
        onC={__VueDX_TypeCheck.internal.first([
          ($event) => {
          call.a['func']($event)
                         ^^^^^^ > 6 at 1711 (44:25)
          }, 
        ])}
`;

exports[`sourcemaps v-if on template + v-slot 1`] = `
import * as __VueDX_TypeCheck from 'vuedx~runtime';
import { defineComponent as __VueDX_defineComponent, GlobalComponents as __VueDX_GlobalComponents } from 'vue';
//#region <script>

const __VueDX__Script_Component = __VueDX_defineComponent({});
//#endregion
//#region <script setup>
 ;const __VueDX__ScriptSetup_scope = __VueDX_TypeCheck.internal.scope(async () => {
const __VueDX__ScriptSetup_props = defineProps({});
const __VueDX__ScriptSetup_emits = ({});
const __VueDX__ScriptSetup_expose = {};
const __VueDX__ScriptSetup_internalProps = {};
const __VueDX__ScriptSetup_Component = __VueDX_defineComponent((_: typeof __VueDX__ScriptSetup_internalProps)=> {});

//#endregion
function __VueDX_RegisterSelf<T>(ctx: T) {
  return { ...ctx, ["Example"]: Example }
}
const __VueDX_ctx = __VueDX_RegisterSelf(new __VueDX__Script_Component())
//#region <template>
/*<vuedx:templateGlobals>*/
const __VueDX__get_identifier_A = () => A;
/*</vuedx:templateGlobals>*/
function __VueDX_render() {
  /*<vuedx:templateGlobals>*/
  let $slots = __VueDX_ctx.$slots
  let foo = __VueDX_ctx.foo
  /*</vuedx:templateGlobals>*/
  /*<vuedx:templateGlobals>*/
  const A = __VueDX_TypeCheck.internal.resolveComponent({} as unknown as __VueDX_GlobalComponents, {} as unknown as JSX.IntrinsicElements, __VueDX_ctx, __VueDX__get_identifier_A(), "A" as const, "A" as const);
  /*</vuedx:templateGlobals>*/
  return (
    <>
      <A
        /*<vuedx:tsx-completions-target/>*/
      >
        {__VueDX_TypeCheck.internal.checkSlots(A, {
          "a": () => {
            return (
              <>
                {$slots.a
                  ? <>
                      {foo}
                    </>
                  : null
                }
              </>
            )
          },
        })}
      </A>
    </>
  )
}
__VueDX_render();
/*<vuedx:diagnosticsIgnore>*/
function __VueDX__slots() {
  /*<vuedx:templateGlobals>*/
  let $slots = __VueDX_ctx.$slots
  let foo = __VueDX_ctx.foo
  /*</vuedx:templateGlobals>*/
  return __VueDX_TypeCheck.internal.union(__VueDX_TypeCheck.internal.flat([
  ]))
}
/*</vuedx:diagnosticsIgnore>*/
const __VueDX__attrs = (() => {
  return __VueDX_TypeCheck.internal.first(__VueDX_TypeCheck.internal.flat([{}]))
})();
//#endregion
return {__VueDX__ScriptSetup_Component, __VueDX__ScriptSetup_props, __VueDX__ScriptSetup_emits, __VueDX__ScriptSetup_expose, __VueDX__attrs, __VueDX__slots, __VueDX_ctx};});
const {__VueDX__ScriptSetup_Component, __VueDX__ScriptSetup_props, __VueDX__ScriptSetup_emits, __VueDX__ScriptSetup_expose, __VueDX__attrs, __VueDX__slots, __VueDX_ctx} = __VueDX__ScriptSetup_scope;
//#region public component definition
const ExamplePublic = null as unknown as new () => typeof __VueDX__ScriptSetup_expose;
export default class Example extends ExamplePublic {
  $props = null as unknown as __VueDX_TypeCheck.internal.MergeAttrs<typeof __VueDX_ctx.$props, typeof __VueDX__attrs>;
  $slots = null as unknown as __VueDX_TypeCheck.internal.Slots<ReturnType<typeof __VueDX__slots>>;
}
//#endregion

`;

exports[`sourcemaps v-on shorthand 1`] = `

    <template>
      <B a="" @a="onStr" />
               ^ > 1 at 31 (2:15)
    </template>
    
----
      <B
        a=""
        onA={__VueDX_TypeCheck.internal.first([
        ^ > 1 at 1436 (35:8)
          onStr, 
        ])}
`;

exports[`sourcemaps v-on shorthand 2`] = `

    <template>
      <B a="" @a="onStr" />
                > 0 at 31 (2:15)
    </template>
    
----
      <B
        a=""
        onA={__VueDX_TypeCheck.internal.first([
        ^ > 1 at 1436 (35:8)
          onStr, 
        ])}
`;

exports[`sourcemaps v-text 1`] = `

    <template>
      <div v-text="a" />
                   ^ > 1 at 35 (2:19)
    </template>
    
----
    <>
      <div
        innerHTML={a}
                   ^ > 1 at 1057 (28:19)
        /*<vuedx:tsx-completions-target/>*/
      />
`;

exports[`sourcemaps v-text 2`] = `

    <template>
      <div v-text="a" />
                   ^ > 1 at 35 (2:19)
    </template>
    
----
    <>
      <div
        innerHTML={a}
                   ^ > 1 at 1057 (28:19)
        /*<vuedx:tsx-completions-target/>*/
      />
`;
